internal/controllers/machineset/machineset_controller.go

```diff
@@ -991,6 +958,82 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
 	return node, nil
 }
 
+func (r *Reconciler) reconcileUnhealthyMachines(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, filteredMachines []*clusterv1.Machine) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// List all unhealthy machines.
+	machinesToRemediate := make([]*clusterv1.Machine, 0, len(filteredMachines))
+	for _, m := range filteredMachines {
+		// filteredMachines contains machines in deleting status to calculate correct status.
+		// skip remediation for those in deleting status.
+		if !m.DeletionTimestamp.IsZero() {
+			continue
+		}
+		if conditions.IsFalse(m, clusterv1.MachineOwnerRemediatedCondition) {
+			machinesToRemediate = append(machinesToRemediate, m)
+		}
+	}
+
+	// If there are no machines to remediate return early.
+	if len(machinesToRemediate) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	preflightChecksPassed := false
```

Comment:
Suggestion to simplify this a bit:
```go
	preflightChecksResult, preflightCheckErrMessage, err := r.runPreflightChecks(ctx, cluster, ms, "Machine Remediation")
	if err != nil {
		// If err is not nil use that as the preflightCheckErrMessage
		preflightCheckErrMessage = err.Error()
	}

	preflightChecksFailed := err != nil || !preflightChecksResult.IsZero()
	if preflightChecksFailed {
		// PreflightChecks did not pass. Update the MachineOwnerRemediated condition on the unhealthy Machines with
		// WaitingForRemediationReason reason.
		var errs []error
		for _, m := range machinesToRemediate {
			patchHelper, err := patch.NewHelper(m, r.Client)
			if err != nil {
				errs = append(errs, errors.Wrapf(err, "failed to create patch helper for Machine %s", klog.KObj(m)))
				continue
			}
			conditions.MarkFalse(m, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, preflightCheckErrMessage)
			if err := patchHelper.Patch(ctx, m); err != nil {
				errs = append(errs, errors.Wrapf(err, "failed to patch Machine %s", klog.KObj(m)))
			}
		}

		if len(errs) > 0 {
			return ctrl.Result{}, errors.Wrapf(kerrors.NewAggregate(errs), "failed to patch unhealthy Machines")
		}
		return preflightChecksResult, nil
	}

	// PreflightChecks passed, so it is safe to remediate unhealthy machines.
	// Remediate unhealthy machines by deleting them.
	var errs []error
	for _, m := range machinesToRemediate {
		log.Info(fmt.Sprintf("Deleting Machine %s because it was marked as unhealthy by the MachineHealthCheck controller", klog.KObj(m)))
		patch := client.MergeFrom(m.DeepCopy())
		if err := r.Client.Delete(ctx, m); err != nil {
			errs = append(errs, errors.Wrapf(err, "failed to delete Machine %s", klog.KObj(m)))
			continue
		}
		conditions.MarkTrue(m, clusterv1.MachineOwnerRemediatedCondition)
		if err := r.Client.Status().Patch(ctx, m, patch); err != nil && !apierrors.IsNotFound(err) {
			errs = append(errs, errors.Wrapf(err, "failed to update status of Machine %s", klog.KObj(m)))
		}
	}

	if len(errs) > 0 {
		return ctrl.Result{}, errors.Wrapf(kerrors.NewAggregate(errs), "failed to delete unhealthy Machines")
	}
	return ctrl.Result{}, nil
}
```
(Please take a look at the diff and let me know if you want me to explain why I changed which part)